crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,459 @@
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetcher: Sync + Send {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its price
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, f64>, PriceEstimationError>;
+}
+
+/// Buffered implementation that implements automatic batching of
+/// native prices requests.
+#[derive(Clone)]
+pub struct BufferedTransport<Inner> {
+    config: Configuration,
+    #[allow(dead_code)]
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+type NativePriceResult = (H160, Result<f64, PriceEstimationError>);
+
+impl<Inner> BufferedTransport<Inner>
+where
+    Inner: NativePriceBatchFetcher + Send + Sync + 'static,
+{
+    /// Maximum capacity of the broadcast channel, the messages are discarded as
+    /// soon as they are sent, so this limit should be enough to hold the
+    /// flow
+    const BROADCAST_CHANNEL_CAPACITY: usize = 50;
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+
+        let (broadcast_sender, _) = broadcast::channel(Self::BROADCAST_CHANNEL_CAPACITY);
+
+        Self::background_worker(
+            inner.clone(),
+            config.clone(),
+            receiver,
+            broadcast_sender.clone(),
+        );
+
+        Self {
+            inner,
+            calls,
+            broadcast_sender,
+            config,
+        }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<H160>,
+        broadcast_sender: broadcast::Sender<NativePriceResult>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if batch.len() != 0 {
+                    let results = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| (token, Ok::<_, PriceEstimationError>(price)))
+                            .collect::<HashMap<_, _>>(),
+                        Err(err) => {
```

Comment:
If we receive an error from the batching call, them all the responses for each token of the batch will have such error.